File: ansible_aisnippet/cache.py
"""In-memory response cache with TTL expiry and LRU-style eviction."""
from __future__ import annotations

import hashlib
import time
from dataclasses import dataclass
from typing import Dict, Optional


@dataclass
class _CacheEntry:
    value: str
    created_at: float
    ttl: int

    def is_expired(self) -> bool:
        return time.time() - self.created_at > self.ttl


class ResponseCache:
    """Thread-safe in-memory cache for AI provider responses.

    Entries are keyed by a hash of ``(provider, system_message, user_message)``
    and expire after *ttl* seconds.  When the cache is full the oldest entry
    is evicted first; expired entries are cleaned up during every ``set``.

    Args:
        ttl:      Time-to-live for each entry in seconds (default 3 600).
        max_size: Maximum number of entries before eviction (default 100).
    """

    def __init__(self, ttl: int = 3600, max_size: int = 100) -> None:
        self.ttl = ttl
        self.max_size = max_size
        self._cache: Dict[str, _CacheEntry] = {}

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def get(
        self, provider: str, system_message: str, user_message: str
    ) -> Optional[str]:
        """Return cached response or ``None`` if missing / expired."""
        key = self._make_key(provider, system_message, user_message)
        entry = self._cache.get(key)
        if entry is None:
            return None
        if entry.is_expired():
            del self._cache[key]
            return None
        return entry.value

    def set(
        self, provider: str, system_message: str, user_message: str, value: str
    ) -> None:
        """Store *value* in the cache, evicting stale / oldest entries if full."""
        key = self._make_key(provider, system_message, user_message)
        self._evict_if_needed()
        self._cache[key] = _CacheEntry(
            value=value, created_at=time.time(), ttl=self.ttl
        )

    def clear(self) -> None:
        """Remove all cached entries."""
        self._cache.clear()

    def __len__(self) -> int:
        return len(self._cache)

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    @staticmethod
    def _make_key(provider: str, system_message: str, user_message: str) -> str:
        content = f"{provider}:{system_message}:{user_message}"
        return hashlib.sha256(content.encode()).hexdigest()

    def _evict_if_needed(self) -> None:
        if len(self._cache) < self.max_size:
            return
        # First remove all expired entries
        expired = [k for k, v in self._cache.items() if v.is_expired()]
        for k in expired:
            del self._cache[k]
        # If still full, remove the oldest live entry
        if len(self._cache) >= self.max_size and self._cache:
            oldest_key = min(self._cache, key=lambda k: self._cache[k].created_at)
            del self._cache[oldest_key]

File: internal/cache/cache.go
// Package cache provides an in-memory response cache with TTL expiry and LRU-style eviction.
package cache

import (
	"crypto/sha256"
	"fmt"
	"sync"
	"time"
)

type entry struct {
	value     string
	createdAt time.Time
	ttl       time.Duration
}

func (e *entry) isExpired() bool {
	return time.Since(e.createdAt) > e.ttl
}

// ResponseCache is a thread-safe in-memory cache for AI provider responses.
//
// Entries are keyed by a hash of (provider, systemMessage, userMessage) and
// expire after ttl. When the cache is full, expired entries are removed first;
// if still full the oldest live entry is evicted.
type ResponseCache struct {
	ttl     time.Duration
	maxSize int
	mu      sync.Mutex
	store   map[string]*entry
	order   []string // insertion order for LRU eviction
}

// New creates a ResponseCache with the given TTL (seconds) and maximum number of entries.
func New(ttlSeconds, maxSize int) *ResponseCache {
	return &ResponseCache{
		ttl:     time.Duration(ttlSeconds) * time.Second,
		maxSize: maxSize,
		store:   make(map[string]*entry),
	}
}

// Get returns the cached response or an empty string and false if missing or expired.
func (c *ResponseCache) Get(provider, systemMessage, userMessage string) (string, bool) {
	key := makeKey(provider, systemMessage, userMessage)
	c.mu.Lock()
	defer c.mu.Unlock()

	e, ok := c.store[key]
	if !ok {
		return "", false
	}
	if e.isExpired() {
		c.deleteKey(key)
		return "", false
	}
	return e.value, true
}

// Set stores a response in the cache, evicting stale or oldest entries if full.
func (c *ResponseCache) Set(provider, systemMessage, userMessage, value string) {
	key := makeKey(provider, systemMessage, userMessage)
	c.mu.Lock()
	defer c.mu.Unlock()

	c.evictIfNeeded()

	if _, exists := c.store[key]; !exists {
		c.order = append(c.order, key)
	}
	c.store[key] = &entry{
		value:     value,
		createdAt: time.Now(),
		ttl:       c.ttl,
	}
}

// Clear removes all cached entries.
func (c *ResponseCache) Clear() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.store = make(map[string]*entry)
	c.order = nil
}

// Len returns the number of entries currently in the cache.
func (c *ResponseCache) Len() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return len(c.store)
}

// makeKey produces a SHA-256 hash key from the provider and messages.
func makeKey(provider, systemMessage, userMessage string) string {
	h := sha256.New()
	fmt.Fprintf(h, "%s:%s:%s", provider, systemMessage, userMessage)
	return fmt.Sprintf("%x", h.Sum(nil))
}

// deleteKey removes a single entry (must be called with mu held).
func (c *ResponseCache) deleteKey(key string) {
	delete(c.store, key)
	for i, k := range c.order {
		if k == key {
			c.order = append(c.order[:i], c.order[i+1:]...)
			break
		}
	}
}

// evictIfNeeded removes expired entries and, if still needed, the oldest live entry.
// Must be called with mu held.
func (c *ResponseCache) evictIfNeeded() {
	if len(c.store) < c.maxSize {
		return
	}
	// Remove all expired entries first.
	for key, e := range c.store {
		if e.isExpired() {
			c.deleteKey(key)
		}
	}
	// If still full, remove the oldest live entry (first in insertion order).
	if len(c.store) >= c.maxSize && len(c.order) > 0 {
		oldest := c.order[0]
		c.deleteKey(oldest)
	}
}
